File: JavaScript Implementation/symbol-table/SymbolTable.js
const TypeSymbol = require('./TypeSymbol');

class SymbolTable {
    constructor(scopeName, scopeLevel, enclosingScope = null) {
        this.symbols = new Map();
        this.scopeName = scopeName;
        this.scopeLevel = scopeLevel;
        this.enclosingScope = enclosingScope;
        this.initBuiltIn();
    }

    initBuiltIn() {
        this.define(TypeSymbol.create('INTEGER'));
        this.define(TypeSymbol.create('CHAR'));
        this.define(TypeSymbol.create('STRING'));
        this.define(TypeSymbol.create('BOOLEAN'));
        this.define(TypeSymbol.create('FLOAT'));
        this.define(TypeSymbol.create('DOUBLE'));
        
        return this;
    }
    
    define(symbol) {
        this.symbols.set(symbol.getName(), symbol);

        return this;
    }

    lookup(name, currentScopeOnly = false) {
        const symbol = this.symbols.get(name);

        // console.log("SYMBOLS: " + this.symbols.toString());
        if(symbol) return symbol;
        if(currentScopeOnly) return null;
        if(this.enclosingScope) return this.enclosingScope.lookup(name);
    }

    toString() {
        let list = '';
        
        for(const[key, value] of this.symbols.entries()) {
            list += `${key}:${value}\n`;
        }

        return `Scope Name: ${this.scopeName}\nScopeLevel: ${this.scopeLevel}\n\nEntries\n${list}`;
    }

    create(scopeName, scopeLevel, enclosingScope) {
        return new this(scopeName, scopeLevel, enclosingScope);
    }
}

module.exports = SymbolTable;

File: Java Implementation/src/symbol_table/SymbolTable.java
package symbol_table;

import java.util.HashMap;

public class SymbolTable {
    private HashMap<String, Symbol> symbols = new HashMap<String, Symbol>();
    private String scopeName;
    private int scopeLevel;
    private Object enclosingScope;

    public SymbolTable(String scopeName, int scopeLevel, Object enclosingScope){
        this.scopeName = scopeName;
        this.scopeLevel = scopeLevel;
        this.enclosingScope = enclosingScope;
        this.symbols = new HashMap<String, Symbol>();

        init();
    }

    public SymbolTable init(){
        this.define(new TypeSymbol("INTEGER", null));
        this.define(new TypeSymbol("BOOLEAN", null));
        this.define(new TypeSymbol("STRING", null));
        this.define(new TypeSymbol("CHAR", null));

        return this;
    }

    public void define(Symbol s){
        this.symbols.put(s.getName(), s);
    }
}